src/features/broker-protection/actions/build-url-transforms.js

```diff
@@ -0,0 +1,158 @@
+import { getStateFromAbbreviation } from '../comparisons/address.js'
+
+/**
+ * @typedef {{url: string} & Record<string, any>} BuildUrlAction
+ * @typedef {Record<string, string|number>} UserData
+ */
+
+/**
+ * Input: { url: 'https://example.com/a/${firstName}-${lastName}', ... }
+ * Output: { url: 'https://example.com/a/John-Smith' }
+ *
+ * @param {BuildUrlAction} action
+ * @param {Record<string, string|number>} userData
+ * @return {{ url: string } | { error: string }}
+ */
+export function transformUrl (action, userData) {
+    const url = new URL(action.url)
+
+    /**
+     * assign the updated pathname + search params
+     */
+    url.search = processSearchParams(url.searchParams, action, userData).toString()
+    url.pathname = processPathname(url.pathname, action, userData)
+
+    /**
+     * Finally, convert back to a full URL
+     */
+    return { url: url.toString() }
+}
+
+/**
+ * These will be applied by default if the key exists in the data.
+ *
+ * @type {Map<string, ((value: string) => string)>}
+ */
+const baseTransforms = new Map([
+    ['firstName', (value) => capitalize(value)],
+    ['lastName', (value) => capitalize(value)],
+    ['state', (value) => value.toLowerCase()],
+    ['city', (value) => capitalize(value)],
+    ['age', (value) => value.toString()]
+])
+
+/**
+ * These are optional transforms, will be applied when key is found in the
+ * variable syntax
+ *
+ * Example, `/a/b/${name|capitalize}` -> applies the `capitalize` transform
+ * to the name field
+ *
+ * @type {Map<string, ((value: string, action: BuildUrlAction) => string)>}
+ */
+const optionalTransforms = new Map([
+    ['hyphenated', (value) => value.split(' ').join('-')],
+    ['capitalize', (value) => capitalize(value)],
+    ['downcase', (value) => value.toLowerCase()],
+    ['upcase', (value) => value.toUpperCase()],
+    ['snakecase', (value) => value.split(' ').join('_')],
+    ['stateFull', (value) => getStateFromAbbreviation(value)],
+    ['ageRange', (value, action) => {
+        if (!action.ageRange) return value
+        const ageNumber = Number(value)
+        // find matching age range
+        const ageRange = action.ageRange.find(range => {
+            const [min, max] = range.split('-')
+            return ageNumber >= Number(min) && ageNumber <= Number(max)
+        })
+        return ageRange || value
+    }]
+])
+
+/**
+ * Take an instance of URLSearchParams and process a new one, with each value
+ * processed
```

Comment:
What kind of processing?

Reply:
oops! typo, will fix. should just be 'produce a new one'